File: Calcolatrice_C++/Model/quadrilatero.cpp
#include "quadrilatero.h"
#include "poligonoconvesso.h"

const unsigned int Quadrilatero::numLati = 4;

Quadrilatero::Quadrilatero(QString s, const vector<Vertice> &v) : PoligonoConvesso(s, v, 4){}

Quadrilatero* Quadrilatero::clone() const
{
    return new Quadrilatero(*this);
}

double Quadrilatero::getPerimetro() const
{
    double per=0;
    for(unsigned int i=0; i<Vertici.size(); ++i)
    {
        if(i+1 == Vertici.size())
            per += distanza(Vertici[i], Vertici[0]);
        else
            per += distanza(Vertici[i], Vertici[i+1]);
    }
    return per;
}

double Quadrilatero::getArea() const
{ // formula di Brahmagupta
    double p = getPerimetro()/2;

    double a = distanza(Vertici[0], Vertici[1]);
    double b = distanza(Vertici[1], Vertici[2]);
    double c = distanza(Vertici[2], Vertici[3]);
    double d = distanza(Vertici[3], Vertici[0]);

    return sqrt((p-a)*(p-b)*(p-c)*(p-d));
}

double Quadrilatero::getSommaAngoli() const
{
    return 360;
}

unsigned int Quadrilatero::getNumLati() const
{
    return numLati;
}

double Quadrilatero::getDiagolale(unsigned int v1) const
{
    if(v1<Vertici.size())
    {
        if(v1==Vertici.size()-1)
            return distanza(Vertici[v1], Vertici[1]);
        else if(v1==Vertici.size()-2)
            return distanza(Vertici[v1], Vertici[0]);
        else
            return distanza(Vertici[v1], Vertici[v1+2]);
    }
    else
        throw std::invalid_argument("Posizioni non valide!");
}

bool Quadrilatero::operator<(const Figura &f) const
{
    return getArea() < f.getArea();
}

bool Quadrilatero::operator>(const Figura &f) const
{
    return getArea() > f.getArea();
}

string Quadrilatero::toString() const
{
    stringstream buffer;
    buffer << *this;
    return buffer.str();
}

ostream &operator<<(ostream &os, const Quadrilatero &q)
{
    q.print(os);
    return os;
}

File: Calcolatrice_Java/src/Quadrilatero.java
package PoligoniConvessi;

import java.util.ArrayList;

import Eccezioni.simpleExc;

import vertice.*;

public class Quadrilatero extends PoligonoConvesso {
	protected static final int numLati=4;
	
	public Quadrilatero(String name, final ArrayList<Vertice> v) throws simpleExc {
		super(name, v, 4);
	}

	protected Quadrilatero(Quadrilatero q) {
		super(q);
	}
	
	@Override
	public double getSommaAngoli() 
	{ return 360; }

	@Override
	public int getNumLati() 
	{ return numLati; }

	@Override
	public double getPerimetro() {
		double per=0;
	    for(int i=0; i<Vertici.size(); ++i)
	    {
	        if(i+1 == Vertici.size())
	            per += Punto.distanza(Vertici.get(i), Vertici.get(0));
	        else
	            per += Punto.distanza(Vertici.get(i), Vertici.get(i+1));
	    }
	    return per;
	}

	@Override
	public double getArea() {
	// formula di Brahmagupta
	    double p = getPerimetro()/2;
	    double a = Punto.distanza(Vertici.get(0), Vertici.get(1));
	    double b = Punto.distanza(Vertici.get(1), Vertici.get(2));
	    double c = Punto.distanza(Vertici.get(2), Vertici.get(3));
	    double d = Punto.distanza(Vertici.get(3), Vertici.get(0));
	    return Math.sqrt((p-a)*(p-b)*(p-c)*(p-d));
	}

	public double getDiagolale(int v1, int v2) throws simpleExc
	{
		if(v1>=0 && v1<Vertici.size() && v2>=0 && v2<Vertici.size()
	            && (v1==v2+2 || v2==v1+2))
	    {
	        if((v1==Vertici.size()-1 && v2==0)
	                || (v1==0 && v2==Vertici.size()-1))
	            return Punto.distanza(Vertici.get(0), Vertici.get(Vertici.size()-1));
	        else
	            return Punto.distanza(Vertici.get(v1), Vertici.get(v2));
	    }
	    else
	        throw new simpleExc("Posizioni non valide!");
	}

	@Override
	public String toString() {
		return super.toString();
	}
	
	@Override
	public boolean equals(Object obj) {
		if(!(obj instanceof Quadrilatero)) 
			return false;
		if(this == obj) 
			return true;
		Quadrilatero p = (Quadrilatero)obj;
		return this.equals(p); 
	}

	@Override
	public Quadrilatero clone() {
		return new Quadrilatero(this);
	}
	
}
